[MRG] TST use global_random_seed in sklearn/ensemble/tests/test_gradient_boosting_loss_functions.py

haochunchang opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Towards #22827 

#### What does this implement/fix? Explain your changes.
* Add `global_random_seed` fixture to 3 tests.
* Change `test_lad_equals_quantiles` seed to 0-99 to cover all possible seed.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

haochunchang committed:
Add global_random_seed to test_multinomial_deviance

haochunchang committed:
Add global_random_seed to test_init_raw_predictions_shapes

haochunchang committed:
Add global_random_seed to test_init_raw_predictions_values

haochunchang committed:
Change seed of test_lad_equals_quantiles to 0-99

glemaitre commented:
Could you change the following test:

```python
@pytest.mark.parametrize("loss", ("log_loss", "exponential"))
def test_classification_synthetic(loss, global_random_seed):
    # Test GradientBoostingClassifier on synthetic dataset used by
    # Hastie et al. in ESLII - Figure 10.9
    X, y = datasets.make_hastie_10_2(n_samples=12000, random_state=global_random_seed)

    X_train, X_test = X[:2000], X[2000:]
    y_train, y_test = y[:2000], y[2000:]

    # Increasing the number of trees should decrease the test error
    common_params = {
        "max_depth": 1,
        "learning_rate": 1.0,
        "loss": loss,
        "random_state": global_random_seed,
    }
    gbrt_100_stumps = GradientBoostingClassifier(n_estimators=100, **common_params)
    gbrt_100_stumps.fit(X_train, y_train)

    gbrt_200_stumps = GradientBoostingClassifier(n_estimators=200, **common_params)
    gbrt_200_stumps.fit(X_train, y_train)

    assert gbrt_100_stumps.score(X_test, y_test) < gbrt_200_stumps.score(X_test, y_test)

    # Decision stumps are better suited for this dataset with a large number of
    # estimators.
    common_params = {
        "n_estimators": 200,
        "learning_rate": 1.0,
        "loss": loss,
        "random_state": global_random_seed,
    }
    gbrt_stumps = GradientBoostingClassifier(max_depth=1, **common_params)
    gbrt_stumps.fit(X_train, y_train)

    gbrt_10_nodes = GradientBoostingClassifier(max_leaf_nodes=10, **common_params)
    gbrt_10_nodes.fit(X_train, y_train)

    assert gbrt_stumps.score(X_test, y_test) > gbrt_10_nodes.score(X_test, y_test)
```

I checked the ESL book and it was not obvious what it was testing. By changing the random state, the test was failing.

So checking, the book is advocating that stumps are more suitable than deeper trees with this dataset (p. 363). So I modified the test accordingly and make sure that it passes all tests with different random state.

glemaitre commented:
You can add the global random seed to `test_regression_dataset`, you will need to modify the MSE threshold from `0.04` to `0.045`

glemaitre commented:
Another test to modify and relaxing the MSE error:

```python
def test_regression_synthetic(global_random_seed):
    # Test on synthetic regression datasets used in Leo Breiman,
    # `Bagging Predictors?. Machine Learning 24(2): 123-140 (1996).
    random_state = check_random_state(global_random_seed)
    regression_params = {
        "n_estimators": 100,
        "max_depth": 4,
        "min_samples_split": 2,
        "learning_rate": 0.1,
        "loss": "squared_error",
    }

    # Friedman1
    X, y = datasets.make_friedman1(n_samples=1200, random_state=random_state, noise=1.0)
    X_train, y_train = X[:200], y[:200]
    X_test, y_test = X[200:], y[200:]

    clf = GradientBoostingRegressor()
    clf.fit(X_train, y_train)
    mse = mean_squared_error(y_test, clf.predict(X_test))
    assert mse < 5.5

    # Friedman2
    X, y = datasets.make_friedman2(n_samples=1200, random_state=random_state)
    X_train, y_train = X[:200], y[:200]
    X_test, y_test = X[200:], y[200:]

    clf = GradientBoostingRegressor(**regression_params)
    clf.fit(X_train, y_train)
    mse = mean_squared_error(y_test, clf.predict(X_test))
    assert mse < 2500.0

    # Friedman3
    X, y = datasets.make_friedman3(n_samples=1200, random_state=random_state)
    X_train, y_train = X[:200], y[:200]
    X_test, y_test = X[200:], y[200:]

    clf = GradientBoostingRegressor(**regression_params)
    clf.fit(X_train, y_train)
    mse = mean_squared_error(y_test, clf.predict(X_test))
    assert mse < 0.025
```

ogrisel commented on sklearn/ensemble/tests/test_gradient_boosting_loss_functions.py:
```diff
@@ -294,7 +294,7 @@ def test_init_raw_predictions_values():
 
 
-@pytest.mark.parametrize("seed", range(5))
+@pytest.mark.parametrize("seed", range(100))
 @pytest.mark.parametrize("alpha", [0.4, 0.5, 0.6])
 def test_lad_equals_quantiles(seed, alpha):
     # Make sure quantile loss with alpha = .5 is equivalent to LAD
```
Please use the `global_random_seed` fixture here as well instead of exhaustive testing at each run.

```suggestion
@pytest.mark.parametrize("alpha", [0.4, 0.5, 0.6])
def test_lad_equals_quantiles(global_random_seed, alpha):
    # Make sure quantile loss with alpha = .5 is equivalent to LAD
```

The line:

```python
        rng = np.random.RandomState(seed)
```

needs to be updated but I cannot do it in the github web interface because it's not close enough to the diff.

haochunchang replied:
Thanks!

haochunchang commented:
> Ups I see that I did not put my comment for the right test file. @haochunchang you can use the comments for the other PR of yours: #23549

No problem, thank you

haochunchang committed:
Use global_random_seed in ensemble/test_gradient_boosting_loss_function.py

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/haochunchang/23559

jeremiedbb committed:
check [all random seeds]
test_multinomial_deviance
test_init_raw_predictions_values
test_lad_equals_quantiles

jeremiedbb reviewed (approved):
Thanks for the PR @haochunchang. LGTM
